src/parser.rs

```diff
@@ -166,51 +155,244 @@ impl<'a> DocParser<'a> {
     }
   }
 
-  /// Parses a module into a list of exported items,
-  /// as well as a list of reexported items which need to be fetched from other modules.
-  pub fn parse_module(
-    &self,
-    specifier: &ModuleSpecifier,
-  ) -> Result<ModuleDoc, DocError> {
-    let module_info = self.get_module_info(specifier)?;
-    let definitions = self.get_doc_nodes_for_module_info(module_info)?;
-    self.collect_diagnostics_for_nodes(&definitions);
-    let reexports = self.get_reexports_for_module(module_info);
-    let module_doc = ModuleDoc {
-      definitions,
-      reexports,
-    };
-    Ok(module_doc)
-  }
-
   fn get_module_info(
     &self,
     specifier: &ModuleSpecifier,
   ) -> Result<ModuleInfoRef, DocError> {
     get_module_info(&self.root_symbol, specifier)
   }
 
-  /// Fetches `file_name` and returns a list of exported items (no reexports).
-  #[cfg(feature = "rust")]
   pub fn parse(
     &self,
+  ) -> Result<IndexMap<ModuleSpecifier, Vec<DocNode>>, DocError> {
+    let mut doc_nodes_by_url = self
+      .specifiers
+      .iter()
+      .map(|specifier| {
+        Ok((
+          specifier.clone(),
+          self.parse_with_reexports_inner(specifier, HashSet::new())?,
+        ))
+      })
+      .collect::<Result<IndexMap<_, _>, DocError>>()?;
+
+    let mut all_locations = doc_nodes_by_url
+      .values()
+      .flatten()
+      .flat_map(|node| {
+        fn walk_rc_nodes(nodes: &[Rc<DocNode>]) -> Vec<Location> {
+          nodes
+            .iter()
+            .flat_map(|node| {
+              if let Some(namespace) = node.namespace_def() {
+                walk_rc_nodes(&namespace.elements)
```

Comment:
Is it possible for someone to create a circular reference by being stupid and cause an inifinte loop here?

Reply:
i'll add a test